lib/remap-lang/src/expression.rs

```diff
@@ -92,6 +97,88 @@ impl ResolveKind {
             Maybe(v) => v.value_kinds(),
         }
     }
+
+    /// Merge two [`ResolveKind`]s, such that the new `ResolveKind` provides the
+    /// most constraint possible resolve kind variant.
+    pub fn merge(&self, other: &Self) -> Self {
+        use ResolveKind::*;
+
+        if let Maybe(kind) = self {
+            let other = match other {
+                Maybe(v) => v,
+                _ => other,
+            };
+
+            return Maybe(Box::new(kind.merge(other)));
+        }
+
+        if let Maybe(kind) = other {
+            return Maybe(Box::new(self.merge(kind)));
+        }
+
+        if self.is_any() || other.is_any() {
+            return Any;
+        }
+
```

Comment:
```rust
match (self, other) {
   (Maybe(kind), Maybe(other)) => return Maybe(Box::new(kind.merge(other)),
   (Maybe(kind), _) => return Maybe(Box::new(kind.merge(other)),
   (_, Maybe(other)) => return Maybe(Box::new(self.merge(other))),
   (Any, _) => return Any,
   (_, Any) => return Any,
   _ => ()
}
```

Maybe a bit of duplication, but I think this makes it clearer what it is trying to do..

Reply:
Ignore this, I see it's all changed..